src/miral/static_display_config.cpp

```diff
@@ -459,3 +475,133 @@ auto miral::StaticDisplayConfig::list_layouts() const -> std::vector<std::string
     return result;
 }
 
+miral::ReloadingYamlFileDisplayConfig::ReloadingYamlFileDisplayConfig(std::string basename) :
+    basename{std::move(basename)}
+{
+}
+
+miral::ReloadingYamlFileDisplayConfig::~ReloadingYamlFileDisplayConfig()
+{
+    if (auto const ml = the_main_loop())
+    {
+        ml->unregister_fd_handler(this);
+    }
+}
+
+void miral::ReloadingYamlFileDisplayConfig::init_auto_reload(mir::Server& server)
+{
+    {
+        std::lock_guard lock{mutex};
+        the_display_configuration_controller_ = server.the_display_configuration_controller();
+        the_main_loop_ = server.the_main_loop();
+    }
+    auto_reload();
+}
+
+void miral::ReloadingYamlFileDisplayConfig::config_path(std::string newpath)
+{
+    std::lock_guard lock{mutex};
+    config_path_ = newpath;
+}
+
+void miral::ReloadingYamlFileDisplayConfig::dump_config(std::function<void(std::ostream&)> const& print_template_config)
+{
+    if (!config_path_)
+    {
+        mir::log_debug("Nowhere to write display configuration template: Neither XDG_CONFIG_HOME or HOME is set");
+    }
+    else
+    {
+        auto const filename = config_path_.value() + "/" + basename;
+
+        if (access(filename.c_str(), F_OK))
+        {
+            std::ofstream out{filename};
+            print_template_config(out);
+
+            mir::log_debug(
+                "%s display configuration template: %s",
+                out ? "Wrote" : "Failed writing",
+                filename.c_str());
+        }
+    }
+
+    YamlFileDisplayConfig::dump_config(print_template_config);
+}
+
+
+auto miral::ReloadingYamlFileDisplayConfig::the_main_loop() const -> std::shared_ptr<mir::MainLoop>
+{
+    std::lock_guard lock{mutex};
+    return the_main_loop_.lock();
+}
+
+auto miral::ReloadingYamlFileDisplayConfig::the_display_configuration_controller() const -> std::shared_ptr<mir::shell::DisplayConfigurationController>
+{
+    std::lock_guard lock{mutex};
+    return the_display_configuration_controller_.lock();
+}
+
+auto miral::ReloadingYamlFileDisplayConfig::inotify_config_path() -> std::optional<mir::Fd>
+{
+    std::lock_guard lock{mutex};
+
+    if (!config_path_) return std::nullopt;
+
+    mir::Fd inotify_fd{inotify_init()};
+
+    if (inotify_fd < 0)
+        BOOST_THROW_EXCEPTION((std::system_error{errno, std::system_category(), "Failed to initialize inotify_fd"}));
+
+    config_path_wd = inotify_add_watch(inotify_fd, config_path_.value().c_str(), IN_CLOSE_WRITE | IN_CREATE | IN_DELETE);
+
+    return inotify_fd;
+}
+
+void miral::ReloadingYamlFileDisplayConfig::auto_reload()
+{
+    if (auto const icf = inotify_config_path())
```

Comment:
Should we not unregister any existing handlers here? I know in the current use it won't happen, but if anyone would call `init_auto_reload()` more than once, we'd end up handling notify's multiple times?

Reply:
Maybe just abort for a logic error?

Reply:
Sure.